design/dpop.md

```diff
@@ -0,0 +1,308 @@
+
+> To request an access token that is bound to a public key using DPoP,
+   the client MUST provide a valid DPoP proof JWT in a "DPoP" header
+   when making an access token request to the authorization server's
+   token endpoint.  This is applicable for all access token requests
+   regardless of grant type (including, for example, the common
+   "authorization\_code" and "refresh\_token" grant types but also
+   extension grants such as the JWT authorization grant \[RFC7523\]).
+
+_([5.5. DPoP Access Token Request][6])_
+
+First and foremost, the `authorization_code` and `refresh_token` grant types must be supported; support for other grant types (ROPC, client credentials, token exchange, extension grants) could be added later.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.TokenEndpoint
+  * processGrantRequest()
+
+#### Introspection Endpoint
+
+For DPoP-bound tokens, token introspection endpoint must return an additional `cnf` claim with the `jkt` member containing JWK thumbprint (a SHA-256 hash).
+
+_([6.2. JWK Thumbprint Confirmation Method in Token Introspection][8])_
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.AccessTokenIntrospectionProvider
+  * introspect()
+
+#### Metadata Endpoint
+
+With DPoP enabled, a new parameter, `dpop_signing_alg_values_supported`, must be included into the authorization server metadata.
+
+_([5.1. Authorization Server Metadata][10])_
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.OIDCWellKnownProvider
+	* getConfig()
+* org.keycloak.protocol.oidc.representations.OIDCConfigurationRepresentation
+
+#### UserInfo Endpoint
+
+The spec doesn’t directly mention the UserInfo endpoint, but it clearly falls into the category of bearer token protected resources:
+
+> The UserInfo Endpoint is an OAuth 2.0 Protected Resource that returns Claims about the authenticated End-User. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication.
+
+_(OpenID Connect Core 1.0,_ [_5.3. UserInfo Endpoint_][11]_)_
+
+Therefore, UserInfo endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.UserInfoEndpoint
+	* issueUserInfo()
+
+#### Logout Endpoint
+
+The `POST` variant of logout invocation uses refresh token, which could be DPoP-bound. Therefore, Logout endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.LogoutEndpoint
+	* logoutToken()
+
+#### Other Endpoints
+
+The following endpoints in Keycloak are public:
+
+* Server Discovery
+* Server JWK set
+* Authorization
+
+The following endpoints don’t use bearer tokens, but rather client credentials:
+
+* Token Introspection
+* Token Revocation
+
+The following endpoints use cookie authentication:
+
+* Check Session iframe
+
+Therefore, no other endpoints should be affected.
+
+#### Client Registration
```

Comment:
Are there any client specific metadata for DPoP? I am not seeing any in the https://tools.ietf.org/id/draft-ietf-oauth-dpop-03.html specification, but maybe I am missing something?